saleor/checkout/utils.py

```diff
@@ -355,39 +359,67 @@ def update_checkout_quantity(checkout):
     checkout.save(update_fields=["quantity"])
 
 
-def add_variant_to_checkout(
+def check_variant_in_stock(
     checkout, variant, quantity=1, replace=False, check_quantity=True
-):
-    """Add a product variant to checkout.
+) -> Tuple[int, Optional[CheckoutLine]]:
+    """Check if a given variant is in stock and return the new quantity + line"""
+    line = checkout.lines.filter(variant=variant).first()
+    line_quantity = 0 if line is None else line.quantity
 
-    The `data` parameter may be used to differentiate between items with
-    different customization options.
-
-    If `replace` is truthy then any previous quantity is discarded instead
-    of added to.
-    """
-    line, _ = checkout.lines.get_or_create(
-        variant=variant, defaults={"quantity": 0, "data": {}}
-    )
-    new_quantity = quantity if replace else (quantity + line.quantity)
+    new_quantity = quantity if replace else (quantity + line_quantity)
 
     if new_quantity < 0:
         raise ValueError(
             "%r is not a valid quantity (results in %r)" % (quantity, new_quantity)
         )
 
+    if new_quantity > 0 and check_quantity:
+        variant.check_quantity(new_quantity)
+
+    return new_quantity, line
+
+
+def update_or_create_checkout_line(
```

Comment:
You are checking line twice. Maybe instead of passing line you can do something like this:
```
def update_or_create_checkout_line(
        checkout, variant, new_quantity
):
    """Updates the quantity of a given checkout line to a given quantity.
    If the line doesn't exist, it will be created."""

    line = checkout.lines.filter(variant=variant).first()
    if line is None:
        line = checkout.lines.create(checkout=checkout, variant=variant, quantity=new_quantity)
    if new_quantity == 0 and line:
         line.delete()
    else:
        ...
```

Reply:
The issue here is that we do not want to create a line to simply delete it. That's why there is a nested import. Maybe it would be more explicit to add `elif quantity: ...` instead of `else` even if the conditions become redundant.

Reply:
I just tried. It's way too ugly and complicated to understand.